lib/ae_mdw/blocks.ex

```diff
@@ -67,6 +67,22 @@ defmodule AeMdw.Blocks do
     end
   end
 
+  @spec fetch_micro_block(State.t(), binary()) :: {:ok, block()} | {:error, Error.t()}
+  def fetch_micro_block(state, hash) do
+    with {:ok, hash, height} <- extract_height(state, :micro, hash) do
+      mbi =
+        hash
+        |> Db.get_micro_blocks()
```

Comment:
Not sure I understand what you mean. `:aec_blocks.prev_hash(block)` returns a hash, so I can't do `:aec_blocks.prev_hash(:aec_blocks.prev_hash(hash))`. I need to use `:aec_db.get_block(hash)` or `:aec_db.get_header(hash)` too to get the prev_hash from it

Reply:
It's because the current `micro_block_walker` defines the unfolding function expected by `unfold(acc(), (acc() -> {element(), acc()} | nil))` in a way that the element accumulated is the `block`:
https://github.com/aeternity/ae_mdw/blob/054007411c87bbaea9deee93db4c88cccbb6aea5/lib/ae_mdw/node/db.ex#L245

On the other hand with a simple recursion the accumulator could be the index counter.

Reply:
The only difference between using a number accumulated and having this stream reduce blocks is that it will keep 1 block in-memory, while the other alternative will keep 1 number in-memory. TBH, I don't think it's worth a separate implementation since the block is quite tiny in size

Reply:
What caught my attention on the first moment was `Enum.count` being used on the blocks. The `unfold` keeps all previous microblocks, including its transactions in memory accumulating them to create a list. So one related factor that popped up when I saw it was the comment on the amount of microblocks, that there might thousands in a single generation (decided by the leader limited only the delay of 3 seconds). On top of this taking a more objective number, thinking only on a 100 TPS and supposing all transactions are `spend_tx` (a small kind of transaction with ~ 1KB) it would allocate and reallocate up to 3 * 60 * 100 * 1000 = 18MB per request. In order to handle 1000 simultaneous requests it would require 18GB only for this function call. As the goal of these new endpoints is to provide lighter ones, I think it's a good opportunity to save some resources.

Reply:
What caught my attention on the first moment was `Enum.count` being used on the blocks. The `unfold` keeps all previous microblocks, including its transactions in memory accumulating them to create a list. So one related factor that popped up when I saw it was the comment on the amount of microblocks, that there might thousands in a single generation (decided by the leader limited only by the delay of 3 seconds). On top of this taking a more objective number, thinking only on a 100 TPS and supposing all transactions are `spend_tx` (a small kind of transaction with ~ 1KB) it would allocate and reallocate up to 3 * 60 * 100 * 1000 = 18MB per request. In order to handle 1000 simultaneous requests it would require 18GB only for this function call. As the goal of these new endpoints is to provide lighter ones, I think it's a good opportunity to save some resources.

Reply:
This is not how streams work, streams are lazy, which is why there will be at most 1 micro block in memory at any given time. Please note that at no point the stream is turned into a list
```elixir
stream
|> Enum.to_list()
|> Enum.count()
```
is very different from
```elixir
stream
|> Enum.count()
```

The former will contain all the elements of the Enum in memory, while the latter will contain at most 1 element of the Enum at any given time

Reply:
While `Stream.map` operates on each element `Enum.map` depends on the whole list. Almost sure that `Enum.count` would behave the same but I am checking it.

Reply:
It's true when there is a `Enum.reverse()` like the `micro_block_walker`:
```
iex(aeternity@localhost)1> Task.async(fn -> Stream.unfold(1_000_000, fn 0 -> nil; n -> {n, n-1} end) |> Enum.reverse() |> Enum.count(); :erlang.process_info(self(), :memory) end) |> Task.await()
{:memory, 22445712}
```